src/render/filters.rs

```diff
@@ -274,6 +278,55 @@ impl ResolveFilter for DefaultIfNoneFilter {
     }
 }
 
+impl ResolveFilter for DateFilter {
+    fn resolve<'t, 'py>(
+        &self,
+        variable: Option<Content<'t, 'py>>,
+        py: Python<'py>,
+        template: TemplateString<'t>,
+        context: &mut Context,
+    ) -> ResolveResult<'t, 'py> {
+        let Some(value) = variable else {
+            return Ok(Some("".as_content()));
+        };
+
+        let fmt = match &self.argument {
+            Some(arg) => arg
+                .resolve(py, template, context, ResolveFailures::Raise)?
+                .expect("missing argument in context should already have raised")
+                .to_py(py),
+            None => {
+                let get_format = GET_FORMAT.import(py, "django.utils.formats", "get_format")?;
+                get_format.call1(("DATE_FORMAT",))?
+            }
+        };
+
+        let value_py = value.to_py(py);
+
+        let is_valid = value_py.is_instance_of::<PyDate>()
+            || value_py.is_instance_of::<PyTime>()
+            || value_py.is_instance_of::<PyDateTime>();
+
+        if !is_valid {
+            return Ok(Some("".as_content()));
+        }
+
+        let date_format_fn = DATE_FORMAT.import(py, "django.utils.dateformat", "format")?;
+
+        let formatted = match date_format_fn.call1((value_py, fmt)) {
+            Ok(res) => res,
+            Err(e) => {
+                if e.is_instance_of::<pyo3::exceptions::PyAttributeError>(py) {
+                    return Ok(Some("".as_content()));
+                }
+                return Err(PyRenderError::PyErr(e));
```

Comment:
I'd like us to use `annotate` to add miette formatting to the error for better debugging:

```suggestion
                return Err(PyRenderError::PyErr(e.annotate(py, self.at, "here", template)));
```

`DateFilter` will need an `at` attribute. See `DefaultFilter` for an example.